### DISCUSSION TITLE
DagFileProcessor re-parses dagfiles every time; not only on mtime change

### DISCUSSION THREAD

**Discussion by dlesco** [apache/airflow#38516] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.6.3

### What happened?

I was checking the airflow-scheduler logs, and I found that it is parsing dagfiles on every loop, not only when the mtime of the dagfile changes.

I noticed when looking into one dagfile that takes 12s to parse, since it produces ten DAGs, each DAG with 122 tasks.  The DAGs are disabled in the Airflow UI, have not run in over a week, and the mtime of the dagfile was a week old.  Since the dags were disabled, the processor was not doing things like processing callbacks.

I examined the Airflow code, and I think I see why it is happening.

In `DagBag`, it has an instance variable of:
```python
        self.file_last_changed: dict[str, datetime] = {}
```

The `collect_dags` method calls the `process_file` method, which compares the file mtime to the saved datetime in this instance variable.  So the DagBag code is written to handle skipping the parsing of the dagfile if the mtime hasn't changed.

However, the problem is in DagFileProcessor.  The `process_file` method does the following to get a reference to the DagBag for this dagfile:

```python
dagbag = DagFileProcessor._get_dagbag(file_path)
```

The implementation of this method is:
```python
    @classmethod
    def _get_dagbag(cls, file_path: str):
        try:
            return DagBag(file_path, include_examples=False)
        except Exception:
            cls.logger().exception("Failed at reloading the DAG file %s", file_path)
            Stats.incr("dag_file_refresh_error", tags={"file_path": file_path})
            raise
```

It then saves the dags to the db, but this `dagbag` reference is not saved in memory anywhere.  Next time the DagFileProcessor runs, it constructs a brand new empty DagBag instance by calling the _get_dagbag class method.

### What you think should happen instead?

Instead, the DagFileProcessor should save the dagbag in memory.

The `__init__` method should define a dictionary named self._dagfile_dagbag.

Then `process_file` could call the following method instead of the `_get_dagbag` classmethod:

```python
    def _get_dagfile_dagbag(self, file_path: str):
        try:
            if file_path in self._dagfile_dagbag:
                self._file_path_dagbag[file_path].collect_dags(
                    dag_folder=file_path,
                    include_examples=False
                )
            else:
                self._file_path_dagbag[file_path] = DagBag(
                    file_path,
                    include_examples=False
                )
            return self._file_path_dagbag[file_path]
        except Exception:
            cls.logger().exception("Failed at reloading the DAG file %s", file_path)
            Stats.incr("dag_file_refresh_error", tags={"file_path": file_path})
            raise
```

This will keep the parsed DagBag in memory, and allow skipping parsing if the mtime hasn't changed.

For multiple DagFileProcessors, each one will have to parse it once to get it cached in memory; but that is much better than parsing every time.

### How to reproduce

One can make a simple dagfile that just uses EmptyOperator for the dag, but has top-level code of time.sleep(5).

Then you can check the airflow-scheduler logs, and see if every parse of the file takes longer than 5 seconds, or if many times parsing it takes less time.  In the DagBag.process_file method, the mtime check and early-exit happens before it calls `_load_modules_from_file`.

### Operating System

PRETTY_NAME="Ubuntu 20.04.6 LTS"

### Versions of Apache Airflow Providers

not relevant, core issue.

### Deployment

Google Cloud Composer

### Deployment details

composer version 2.4.6, airflow 2.6.3

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Yes. That's by design and expected - and https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#min-file-process-interval parameter controls how often it happens. 

There are various reasonf for that- 1) Parsing python file might produce different DAG over time (for example when it imports or reads other data 2) dag file processor handles some callbacks and needs to have parsed DAG for that in other toexecute callback methods. By regularly reparsing all the DAGs, it allows Airflow to handle a lot of DAGs without keeping them all parsed and loaded in memory. 

This is a basic feature of Airlfow and changing that would require quite a heavy change in the way how Airflow works. And no - we are 
 not going to optimise it most likely to subset of cases that do not need either of the 2 - because it adds complexity and makes it difficult to handle more complex cases. 

Converting it into a discussion, if more discussion needed - but before you respond - I recommend you to look for similar keywords in past discussions and issues where probably a lot of arguments were raised and questions answered.

**Reply by potiuk** [reply]
> Instead, the DagFileProcessor should save the dagbag in memory.

BTW. That is precisely what we do not want to do. You can optimize for different things. Airflow optimizes for applying dynamic changes to DAGs frequently (even if they are coming from external factors) and for keeping memory footprint relatively low. What you propose is to prefer higher memory usage and less dynamism, not impacted by external factors changing. Those are opposite optimisation goals.

**Reply by dlesco** [reply]
I see that point 1 argument as a valid argument.  That could actually be the case with our setup: our dags read yaml files in the dag folder for configuration.

I think that the documentation on min_file_process_interval can be improved, perhaps to something like this:

> Number of seconds after which a DAG file is parsed. The DAG file is parsed every min_file_process_interval number of seconds. Updates to DAGs are reflected after this interval. Keeping this number low will increase CPU usage.  This is also the interval that callbacks are performed for the DAG: callbacks such as SLA miss callbacks, on_failure_callbacks, on_success_callbacks, and task callbacks.  This setting does not affect how often task instances are scheduled.

I put the last sentence there, because somehow I had in my head that min_file_process_interval affected task instance scheduling, even though it doesn't.  I don't know why that idea was in my head, even though it's not true.  So, I'm thinking it may help other people to include that information there.

**Reply by potiuk** [reply]
> I put the last sentence there, because somehow I had in my head that min_file_process_interval affected task instance scheduling, even though it doesn't. I don't know why that idea was in my head, even though it's not true. So, I'm thinking it may help other people to include that information there.

Absolutely. Feel absolutely free to submit a PR with proposing update to the description. The source of the configuration description is stored in https://github.com/apache/airflow/blob/main/airflow/config_templates/config.yml , and you -  similarly to > 2800 other contributors can submit a change there. Coming from someone who got confused it's even more valuable than maintainers adding those with their assumptions in their heads. Now that you know you shoudl be able to explain it in the way that is best for users like you who got confused. And opening PR gives you a chance to become a contributor and get other maintainers to chime in and maybe even improve your proposal.
